tensorflow/java/src/gen/cc/op_parser.h

```diff
@@ -0,0 +1,137 @@
+
+#ifndef TENSORFLOW_JAVA_SRC_GEN_CC_OP_PARSER_H_
+#define TENSORFLOW_JAVA_SRC_GEN_CC_OP_PARSER_H_
+
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "tensorflow/core/framework/op_def.pb.h"
+#include "tensorflow/core/framework/api_def.pb.h"
+#include "tensorflow/java/src/gen/cc/java_defs.h"
+
+namespace tensorflow {
+namespace java {
+
+// Specification of a TensorFlow operation to generate.
+//
+// This is the result of an operation definition parsing, see OpParser::Parse().
+class OpSpec {
+ public:
+  class Endpoint {
+   public:
+    Endpoint(const Type& type, const Javadoc& javadoc)
+      : type_(type), javadoc_(javadoc) {}
+    const Type& type() const { return type_; }
+    const Javadoc& javadoc() const { return javadoc_; }
+
+   private:
+    Type type_;
+    Javadoc javadoc_;
+  };
+
+  class Operand {
+   public:
+    Operand(const string& graph_name, const Variable& var,
+        const Type& data_type, const string& description, bool iterable)
+     : graph_name_(graph_name), var_(var), data_type_(data_type),
+       description_(description), iterable_(iterable) {}
+    const string& graph_name() const { return graph_name_; }
+    const Variable& var() const { return var_; }
+    Variable* var_ptr() { return &var_; }
+    const Type& data_type() const { return data_type_; }
+    const string& description() const { return description_; }
+    bool iterable() const { return iterable_; }
+
+   private:
+    string graph_name_;
+    Variable var_;
+    Type data_type_;
+    string description_;
+    bool iterable_;
+  };
+
+  explicit OpSpec(const string& graph_name) : graph_name_(graph_name) {}
+  const string& graph_name() const { return graph_name_; }
+  const std::vector<Endpoint> endpoints() const { return endpoints_; }
+  void add_endpoint(const Type& type, const Javadoc& javadoc) {
+    endpoints_.push_back(Endpoint(type, javadoc));
+  }
+  const std::vector<Operand>& inputs() const { return inputs_; }
+  void add_input(const Operand& input) {
+    inputs_.push_back(input);
+  }
+  const std::vector<Operand>& outputs() const { return outputs_; }
+  void add_output(const Operand& output) {
+    outputs_.push_back(output);
+  }
+  const std::vector<Operand>& attributes() const { return attributes_; }
+  void add_attribute(const Operand& attribute) {
+    attributes_.push_back(attribute);
+  }
+  const std::vector<Operand>& options() const { return options_; }
+  void add_option(const Operand& option) {
+    options_.push_back(option);
+  }
+
+ private:
+  string graph_name_;
+  std::vector<Endpoint> endpoints_;
+  std::vector<Operand> inputs_;
+  std::vector<Operand> outputs_;
+  std::vector<Operand> attributes_;
+  std::vector<Operand> options_;
+};
+
+// A parser of ops proto definitions.
+//
+// This object parses the definition and the api of an TensorFlow operation to
+// produce a specification that can be used for Java source code rendering.
+class OpParser {
```

Comment:
Mmh, interesting... I only created a class because I did not want to carry the data shared by helper functions in all their signatures (`visited_attrs_`, `next_generic_`, etc). 

But I can check if that can be part of the `OpSpec` class directly instead.

Reply:
If you do need a class for that, I'd recommend hiding that as an implementation detail in an anonymous namespace in the `.cc` file and keep the header file more focused.